File: results/0.5/Network-Time-Protocol-Version-4/network-time-protocol-version-4-claude-3-5-haiku-20241022.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <hammer/hammer.h>

typedef struct {
    uint8_t leap_indicator;
    uint8_t version;
    uint8_t mode;
    uint8_t stratum;
    int8_t poll_interval;
    int8_t precision;
    double root_delay;
    double root_dispersion;
    char reference_id[5];
    uint64_t reference_timestamp;
    uint64_t origin_timestamp;
    uint64_t receive_timestamp;
    uint64_t transmit_timestamp;
} NTPPacket;

static HParser* ntp_parser;

static HParseResult* parse_ntp_packet(void* data, size_t len) {
    return h_parse(ntp_parser, data, len);
}

static void print_ntp_packet(NTPPacket* packet) {
    printf("NTP Packet Details:\n");
    printf("Leap Indicator: %d\n", packet->leap_indicator);
    printf("Version: %d\n", packet->version);
    printf("Mode: %d\n", packet->mode);
    printf("Stratum: %d\n", packet->stratum);
    printf("Poll Interval: %d\n", packet->poll_interval);
    printf("Precision: %d\n", packet->precision);
    printf("Reference ID: %s\n", packet->reference_id);
}

void initialize_ntp_parser() {
    HParser* leap_indicator = h_bits(2, false);
    HParser* version = h_bits(3, false);
    HParser* mode = h_bits(3, false);
    HParser* stratum = h_bits(8, false);
    HParser* poll = h_int8();
    HParser* precision = h_int8();
    HParser* root_delay = h_float_be();
    HParser* root_dispersion = h_float_be();
    HParser* reference_id = h_sequence(
        h_bits(8, false),
        h_bits(8, false),
        h_bits(8, false),
        h_bits(8, false),
        NULL
    );
    HParser* timestamp = h_bits(64, false);

    ntp_parser = h_sequence(
        leap_indicator,
        version,
        mode,
        stratum,
        poll,
        precision,
        root_delay,
        root_dispersion,
        reference_id,
        timestamp,
        timestamp,
        timestamp,
        timestamp,
        NULL
    );
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <ntp_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }

    size_t read_size = fread(buffer, 1, file_size, file);
    fclose(file);

    if (read_size != file_size) {
        perror("File read error");
        free(buffer);
        return 1;
    }

    initialize_ntp_parser();

    HParseResult* result = parse_ntp_packet(buffer, read_size);
    if (result && result->ast) {
        NTPPacket packet;
        const HParsedToken* tokens = result->ast;

        packet.leap_indicator = tokens[0].uint;
        packet.version = tokens[1].uint;
        packet.mode = tokens[2].uint;
        packet.stratum = tokens[3].uint;
        packet.poll_interval = tokens[4].sint;
        packet.precision = tokens[5].sint;
        packet.root_delay = tokens[6].flt;
        packet.root_dispersion = tokens[7].flt;

        const HCountedArray* ref_id_tokens = tokens[8].seq;
        for (int i = 0; i < 4; i++) {
            packet.reference_id[i] = ref_id_tokens->elements[i].uint;
        }
        packet.reference_id[4] = '\0';

        packet.reference_timestamp = tokens[9].uint;
        packet.origin_timestamp = tokens[10].uint;
        packet.receive_timestamp = tokens[11].uint;
        packet.transmit_timestamp = tokens[12].uint;

        print_ntp_packet(&packet);
        h_parse_result_free(result);
    } else {
        fprintf(stderr, "Parsing failed\n");
    }

    free(buffer);
    return 0;
}

File: results/0.5/Network-Time-Protocol-Version-4/network-time-protocol-version-4-claude-3-5-haiku-20241022.rs
use nom::{
    bits::streaming::take as take_bits,
    bytes::complete::take,
    multi::many0,
    number::complete::{be_f64, be_u16, be_u32, be_u8},
    sequence::tuple,
    IResult,
};
use std::env;
use std::fs;

#[derive(Debug)]
struct NtpPacket {
    leap_indicator: u8,
    version: u8,
    mode: u8,
    stratum: u8,
    poll: i8,
    precision: i8,
    root_delay: f64,
    root_dispersion: f64,
    reference_id: u32,
    reference_timestamp: f64,
    originate_timestamp: f64,
    receive_timestamp: f64,
    transmit_timestamp: f64,
    extensions: Vec<NtpExtension>,
}

#[derive(Debug)]
struct NtpExtension {
    field_type: u16,
    field_length: u16,
    data: Vec<u8>,
}

fn parse_ntp_header(input: &[u8]) -> IResult<&[u8], NtpPacket> {
    let (input, first_byte) = be_u8(input)?;
    let leap_indicator = (first_byte & 0b11000000) >> 6;
    let version = (first_byte & 0b00111000) >> 3;
    let mode = first_byte & 0b00000111;

    let (input, (
        stratum,
        poll,
        precision,
        root_delay,
        root_dispersion,
        reference_id,
        reference_timestamp,
        originate_timestamp,
        receive_timestamp,
        transmit_timestamp,
    )) = tuple((
        be_u8,
        be_u8,
        be_u8,
        be_f64,
        be_f64,
        be_u32,
        be_f64,
        be_f64,
        be_f64,
        be_f64,
    ))(input)?;

    let (input, extensions) = parse_extensions(input)?;

    Ok((input, NtpPacket {
        leap_indicator,
        version,
        mode,
        stratum,
        poll: poll as i8,
        precision: precision as i8,
        root_delay,
        root_dispersion,
        reference_id,
        reference_timestamp,
        originate_timestamp,
        receive_timestamp,
        transmit_timestamp,
        extensions,
    }))
}

fn parse_extensions(input: &[u8]) -> IResult<&[u8], Vec<NtpExtension>> {
    many0(parse_extension)(input)
}

fn parse_extension(input: &[u8]) -> IResult<&[u8], NtpExtension> {
    let (input, field_type) = be_u16(input)?;
    let (input, field_length) = be_u16(input)?;
    let (input, data) = take(field_length)(input)?;

    Ok((input, NtpExtension {
        field_type,
        field_length,
        data: data.to_vec(),
    }))
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <ntp_file>", args[0]);
        std::process::exit(1);
    }

    let file_contents = fs::read(&args[1])?;
    match parse_ntp_header(&file_contents) {
        Ok((_, packet)) => {
            println!("NTP Packet: {:?}", packet);
            Ok(())
        }
        Err(e) => {
            eprintln!("Parsing error: {:?}", e);
            std::process::exit(1)
        }
    }
}